ros/python/py_pubsub/py_pubsub/spatial_mapper.py

```diff
@@ -0,0 +1,445 @@
+        self._det_3d_topic = self.declare_parameter("det_3d_topic", "ObjectDetections3d").get_parameter_value().string_value
+
+        log = self.get_logger()
+        log.info(f"Spatial map topic: {self._spatial_map_topic}")
+        log.info(f"Detection topic: {self._det_topic}")
+        log.info(f"Detection3d topic: {self._det_3d_topic}")
+        log.info(f"Pose topic: {self._headset_pose_topic}")
+
+        self._spatial_mesh_subscription = self.create_subscription(
+            SpatialMesh,
+            self._spatial_map_topic,
+            self.spatial_map_callback,
+            100)
+
+        self._detection_subscription = self.create_subscription(
+            ObjectDetection2dSet,
+            self._det_topic,
+            self.detection_callback,
+            100
+        )
+
+        self._headset_pose_subscription = self.create_subscription(
+            HeadsetPoseData,
+            self._headset_pose_topic,
+            self.headset_pose_callback,
+            100
+        )
+
+        self._object_3d_publisher = self.create_publisher(
+            ObjectDetection3dSet,
+            self._det_3d_topic,
+            1
+        )
+
+        self.frames_recvd = 0
+        self.prev_time = -1
+        self.meshes = {}
+
+        self.scene = trimesh.Scene()
+
+        self.poses = []
+
+        # TODO: get this from a ROS message?
+        self.image_height = 1080
+        self.image_width = 1920
+
+        log = self.get_logger()
+        #log.set_level(10)
+
+
+    def spatial_map_callback(self, msg):
+        log = self.get_logger()
+        self.frames_recvd += 1
+
+        # extract the vertices into np arrays
+        # NOTE: negating the x-component because Unity uses a left-handed
+        # coordinate system and trimesh uses a right-handed coordinate system
+        vertices = np.array([])
+        for v in msg.mesh.vertices:
+            if vertices.size == 0:
+                vertices = np.array([-v.x, v.y, v.z])
+            else:
+                vertices = np.vstack([vertices, np.array([-v.x, v.y, v.z])])
+
+        # extract the triangles into np arrays
+        triangles = np.array([])
+        for t in msg.mesh.triangles:
+            if triangles.size == 0:
+                triangles = np.array([t.vertex_indices[0],
+                                      t.vertex_indices[1],
+                                      t.vertex_indices[2]
+                                     ])
+            else:
+                triangles = np.vstack([triangles,
+                                       np.array([t.vertex_indices[0],
+                                                 t.vertex_indices[1],
+                                                 t.vertex_indices[2]
+                                                ])])
+
+        if msg.removal:
+            log.debug("Got a removal!")
+            # TODO: remove from the spatial map
```

Comment:
Do you intend to add removal with this PR, or is this still a punt for later? Are we prone to "leaking" mesh (accumulation w/o removal) without this removal functionality?

Reply:
I was going to wait on it for now, but can add it if we need. The reason I hadn't added that logic yet is because I've never actually seen the `OnMeshRemoval` callback called in the Unity app.